cvat/apps/engine/cache.py

```diff
@@ -65,63 +70,166 @@
+        cls,
+        key: str,
+        create_callback: Callable[..., DataWithMime],
+        *args: Any,
+        cache_item_ttl: int = None,
+        **kwargs: Any,
+    ) -> DataWithMime:
+        item_data = create_callback(*args, **kwargs)
+        item_data_bytes = item_data[0].getvalue()
+        item = (item_data[0], item_data[1], cls._get_checksum(item_data_bytes))
+        if item_data_bytes:
+            cache = cls._cache()
+            cache.set(key, item, timeout=cache_item_ttl or cache.default_timeout)
+
+        return item
+
+    def _wait_for_rq_job(self, rq_job: rq.job.Job) -> bool:
+        retries = self._CHUNK_CREATE_TIMEOUT // self._SLEEP_TIMEOUT or 1
+        while retries > 0:
+            job_status = rq_job.get_status()
+            if job_status in ("finished",):
+                return True
+            elif job_status in ("failed",):
+                job_meta = rq_job.get_meta()
+                exc_type = job_meta.get(RQJobMetaField.EXCEPTION_TYPE, Exception)
+                exc_args = job_meta.get(RQJobMetaField.EXCEPTION_ARGS, ("Cannot create chunk",))
+                raise exc_type(*exc_args)
+
+            time.sleep(self._SLEEP_TIMEOUT)
+            retries -= 1
+
+        return False
+
+    def _create_cache_item(
+            self,
+            key: str,
+            create_callback: Callable[..., DataWithMime],
+            *args: Any,
+            cache_item_ttl: int = None,
+            **kwargs: Any,
+        ) -> _CacheItem:
+
+        slogger.glob.info(f"Starting to prepare chunk: key {key}")
+        if self._is_run_inside_rq():
+            item = self._create_and_set_cache_item(
+                key,
+                create_callback,
+                *args,
+                cache_item_ttl=cache_item_ttl,
+                **kwargs,
+            )
         else:
-            # compare checksum
-            item_data = item[0].getbuffer() if isinstance(item[0], io.BytesIO) else item[0]
-            item_checksum = item[2] if len(item) == 3 else None
-            if item_checksum != self._get_checksum(item_data):
-                slogger.glob.info(f"Recreating cache item {key} due to checksum mismatch")
-                item = create_item()
+            queue = self._get_queue()
+            rq_id = self._make_queue_job_id_base(key)
+            rq_job = queue.fetch_job(rq_id)
+
+            if not rq_job:
+                rq_job = queue.enqueue(
+                    self._drop_return_value,
+                    self._create_and_set_cache_item,
+                    key,
+                    create_callback,
+                    *args,
+                    cache_item_ttl=cache_item_ttl,
+                    **kwargs,
+                    job_id=rq_id,
+                    result_ttl=self._RQ_JOB_RESULT_TTL,
+                    failure_ttl=self._RQ_JOB_FAILURE_TTL,
+                )
+
+            if self._wait_for_rq_job(rq_job):
+                item = self._get_cache_item(key)
```

Comment:
The returned entry here can be different from what was written, consider adding some checks about it. It's not always a bad thing (e.g. 2 workers generated the same data), but sometimes the data can be different from what's expected (e.g., if the chunk is requested, and it was modified by honeypot replacement).